src/test/ui/intrinsics/panic-uninitialized-zeroed-strict.rs

```diff
@@ -0,0 +1,95 @@
+// run-pass
+// needs-unwind
+// ignore-wasm32-bare compiled with panic=abort by default
+// revisions: mir thir
+// [thir]compile-flags: -Zthir-unsafeck -Zstrict-init-checks
+// [mir]compile-flags: -Zstrict-init-checks
+// ignore-tidy-linelength
+
+// This test checks panic emitted from `mem::{uninitialized,zeroed}`.
+
+#![feature(never_type, arbitrary_enum_discriminant)]
+#![allow(deprecated, invalid_value)]
+
+use std::{
+    mem::{self, MaybeUninit, ManuallyDrop},
+    ptr::NonNull,
+    panic,
+    num,
+};
+
+enum Bar {}
+
+#[allow(dead_code)]
+enum OneVariant { Variant(i32) }
+
+#[allow(dead_code, non_camel_case_types)]
+enum OneVariant_NonZero {
+    Variant(i32, i32, num::NonZeroI32),
+    DeadVariant(Bar),
+}
+
+// An `Aggregate` abi enum where 0 is not a valid discriminant.
+#[allow(dead_code)]
+#[repr(i32)]
+enum NoNullVariant {
+    Variant1(i32, i32) = 1,
+    Variant2(i32, i32) = 2,
+}
+
+// An enum with ScalarPair layout
+#[allow(dead_code)]
+enum LR {
+    Left(i64),
+    Right(i64),
+}
+#[allow(dead_code, non_camel_case_types)]
+enum LR_NonZero {
+    Left(num::NonZeroI64),
+    Right(num::NonZeroI64),
+}
+
+
+
+fn test_panic_msg<T>(op: impl (FnOnce() -> T) + panic::UnwindSafe, msg: &str) {
+    let err = panic::catch_unwind(op).err();
+    assert_eq!(
+        err.as_ref().and_then(|a| a.downcast_ref::<&str>()),
+        Some(&msg)
+    );
+}
+
+fn main() {
+    unsafe {
+        test_panic_msg(
+            || mem::uninitialized::<[u32; 1]>(),
+            "attempted to leave type `[u32; 1]` uninitialized, which is invalid"
+        );
+
+        test_panic_msg(
+            || mem::uninitialized::<i32>(),
+            "attempted to leave type `i32` uninitialized, which is invalid"
+        );
+
+        test_panic_msg(
+            || mem::uninitialized::<*const ()>(),
+            "attempted to leave type `*const ()` uninitialized, which is invalid"
+        );
+
+        test_panic_msg(
+            || mem::zeroed::<[fn(); 1]>(),
+            "attempted to zero-initialize type `[fn(); 1]`, which is invalid"
+        );
```

Comment:
This errors even without the strict flag, right?

Reply:
It does not, because it's in an array.

We currently just give up for arrays, the type in them is completely unchecked, you can make an uninit `[String; 1]` just fine if you want. (Except for uninhabited types, you *can't* make a `[!; 1]`)

Reply:
Ah right.